Peer snapshot documentation (#2043)
hyperledger/fabric: docs/source/peer_ledger_snapshot.md

# Taking ledger snapshots and using them to join channels

> For a peer to contribute to a channel, it must have a copy of the blockchain and maintain a copy of the state database. This state database tracks what's known as the "world state", which represents the current value of all of the keys on the ledger (who owns a particular car, for example). There are two ways for a peer to get a copy of the ledger.

Reviewer: For a peer to participate in a channel, it must maintain minimum ledger data for the channel that is essential for the peer to act as an endorser and as a committer. The "world state" is one such data item. The "world state" is maintained in the state database in the peer and represents the latest values of all of the keys on the ledger (who owns a particular car, for example), as of the last committed block to the channel by the peer. There are two ways for a peer to join a channel and get a copy of the ledger data.

Author: Made a few edits here, though not exactly your language.

Reviewer: Sure, LGTM

1. Pull the ledger blocks from the ordering service, starting with the initial configuration block (known as the "genesis block"), and continue processing blocks until the latest block that has been written to the ledger is reached.

> 2. Join the channel from a "snapshot", which contains the world state but not the individual blocks.

Reviewer: 1. Join the channel from the initial configuration block (known as the "genesis block"), and thereafter, continue pulling the further blocks from the ordering service and processing these blocks locally.

> 2. which contains the world state 

which contains the minimum ledger data as of a particular block number

> While the first method represents a more comprehensive way of joining a channel, because of the size of established channels (which can reach many thousands of blocks), it can take a long time for peers to finish joining a channel. Peers that join a channel this way must also store every block since the creation of the channel, increasing storage costs for an organization.

Reviewer: > to finish joining a channel

 to finish processing all the blocks already committed to the channel

In this topic, we'll describe the process for joining a peer to a channel using a snapshot.

## Limitations

> While joining a peer to a channel from a snapshot (also known as a "checkpoint") will take less time and saves on storage costs compared to processing and storing every block on the ledger, there are a few limitations to consider:

Reviewer: > from a snapshot (also known as a "checkpoint") will take less time

from a snapshot takes lesser time (Better to use only the term snapshot throughout)

Author: We've learned from other docs that it can be good to mention these other alternative names for things at least once, so that people who search for it can find it.

Reviewer: OK, sure

> * It is not possible for a peer that joins from a snapshot to query blocks taken before the ledger height of the snapshot. If the snapshot the peer uses to join a channel from was taken at block 1000, for example, none of the blocks between 1-999 can be queried. Applications attempting to query for the data in these blocks will have to target a peer that contains the relevant block. Because of this, it is likely that all organizations will want to keep at least one peer that has all of the historical data.

Reviewer: >to query blocks taken before

to serve the queries for the blocks that were committed before

Author: I don't think "serve the queries" is helpful language here. I did add "that were committed" after the word "blocks" here, however.

Reviewer: Sure, agree.

* While endorsements will continue and queries can be submitted, all peers taking a snapshot at a particular height will not commit blocks on the channel while the snapshot is being generated. Because taking a snapshot is a resource-intensive operation, the peer might also be slow to endorse transactions or commit blocks on other channels. For these reasons, it is anticipated that snapshots will only be taken when necessary (for example, when a new peer is joining a channel or when organizations want to verify that no ledger forks have occurred). As long as every organization taking a snapshot has at least one peer that is not taking a snapshot, the operation of your organization and your channels should not be affected.

> * Because the private data between organizations in a channel is likely to be at least somewhat different, private data is not included in the snapshot. Peers that join a channel using a snapshot will discover the collections it is a member of and pull the relevant private data from those peers directly.

Reviewer: >from those peers directly

from the existing member peers directly

* **The importance of querying old data**. Some channels and use cases might call for querying old blocks regularly, while in other channels it might only be necessary to know the current state. This must be balanced by the server costs experienced by your organization.

## Overview

Snapshots can be used by organizations that already have peers on a channel or by organizations new to a channel. Whatever the use case, the process is largely the same.

If the organization is new to the channel, the organizations currently joined to the channel must:

1. **Schedule a snapshot**. These snapshots must be taken at **exactly the same ledger height** on each peer. This will allow the new peer organization to evaluate the snapshots to make sure they contain the same data. This ledger height must be equal or higher than the current ledger height. They cannot be taken from a lower ledger height. If you attempt to schedule a snapshot at a height lower than the current height you will get an error. Note that a peer that already used a snapshot to join a channel can also be used to take a snapshot.

> 2. **When the ledger height is reached, the snapshot is taken by the peer**. The snapshot is comprised of folders that include the public state, hashes of private state, transaction IDs, and the collection config history. A file containing metadata relating to these files is also included. For more information, check out [contents of a snapshot](#contents-of-a-snapshot).

Reviewer: >comprised of folders that include the 

comprised of a folder with the files that contain

> 2. **The snapshot is sent to the new organization**. This must be completed out of band. Because snapshot files are not compressed, it is likely that peer administrators will want to compress these files before sending them.

Reviewer: The following can be added here...

In a typical scenario, the administrator will receive the snapshot from any of the existing organizations but would want to receive the snapshot metadata from more than one organizations in order to verify the snapshot received.

The new organization will then:

> 1. **Evaluate the snapshots from the other organizations to ensure they contain the same data**. An administrator of the peer organization attempting to use the snapshot to join the peer to the channel should independently compute the hashes of the snapshot files and match these with the hashes present in the metadata file.

Reviewer: The following can be added here...

In addition, the administrator may want to match the metadata files from more than one organizations, based on the trust model within the consortium. In some scenarios, the administrator may want the administrators of other organizations to sign the metadata file for its record.

> 2. **Join the peer to the channel using the snapshot**. When the peer has finished joining the channel using the snapshot, it will begin pulling private data according to the collections it member of.

Reviewer: >it member of

it is member of

> 3. **Verify the peer has been join the channel successfully**. For more information, check out [Joining a channel using a snapshot](#joining-a-channel-using-a-snapshot).

Reviewer 2: Typo

**Verify the peer has joined the channel successfully**

If an organization that is already joined to the channel wants to join a new peer using a snapshot, it might decide to skip the process of having other organizations take snapshots and evaluate them, though it is a best practice for an organization to periodically take snapshots of all its peers and compare them to ensure the no ledger forks have occurred. In that case, the organization can take a snapshot immediately and then use the snapshot to join the new peer to the channel.

> **Note: snapshots can also be used to verify that the public state between peers is identical, even if no new peer will use the snapshot to join the channel. This can be useful in identifying if a ledger fork has occurred.**

Reviewer 2: Would it add value to put something along the lines of how snapshots can help identify a ledger fork?  Is this referring to commit hashes, or would other meta data be used, etc?

Reviewer: Yes, indeed. We can add here that the value of the field `snapshot_hash` in the file `_snapshot_additional_metadata.json` in the snapshots generated across peers should be the same.

## Taking a snapshot

For the full list of snapshot-related commands, check out [`peer snapshot` commands](./commands/peersnapshot.html).

Before taking a snapshot, it is a best practice to confirm the current ledger height by issuing a command similar to:

```
peer channel getinfo -c <name of channel>
```

You will see a response similar to:

```
2018-02-25 15:15:44.135 UTC [channelCmd] InitCmdFactory -> INFO 003 Endorser and orderer connections initialized
Blockchain info: {"height":970,"currentBlockHash":"JgK9lcaPUNmFb5Mp1qe1SVMsx3o/22Ct4+n5tejcXCw=","previousBlockHash":"f8lZXoAn3gF86zrFq7L1DzW2aKuabH9Ow6SIE5Y04a4="}
2018-02-25 15:15:44.139 UTC [main] main -> INFO 006 Exiting.....

> ```

Reviewer: I don't think that it's a good idea to include the logs, as they may change over time. Just keeping the middle line should suffice

In this example, the ledger height is `970`.

A snapshot request can be submitted by issuing a command similar to:

```
peer snapshot submitrequest -C <name of channel> -b <ledger height where snapshot will be taken> --peerAddress <address of peer> --tlsRootCertFile <path to root certificate of the TLS CA>
```

For example:

```
peer snapshot submitrequest -C testchannel -b 1000 --peerAddress 127.0.0.1:22509 --tlsRootCertFile tls/cert.pem
```

> **If you give a ledger height of `0`, the snapshot is taken immediately. This is useful for cases when an organization is interested in backing up its own ledger data and does not intend to share the data with another organization. Do not take these "immediate" snapshots in cases when snapshots will be evaluated from multiple peers, as it increases the likelihood that the snapshots will be taken at different ledger heights.**

Reviewer: >backing up its own ledger data

generating snapshot for internal consumption only  (to be consistent what is mentioned above that they should not use this feature as a replacement of backup)

> **If you give a ledger height of `0`, the snapshot is taken immediately. This is useful for cases when an organization is interested in backing up its own ledger data and does not intend to share the data with another organization. Do not take these "immediate" snapshots in cases when snapshots will be evaluated from multiple peers, as it increases the likelihood that the snapshots will be taken at different ledger heights.**

Reviewer 2: I found this section a little confusing - if the snapshot does not have enough data to be used for queries, how does it have enough data to make for a good backup?

Reviewer: Yes, I made a suggestion in the similar lines in this [comment](https://github.com/hyperledger/fabric/pull/2043#discussion_r512318297)

If the request is successful, you will see a `Snapshot request submitted successfully` message.

You can list the pending snapshots by issuing a command similar to:

```
peer snapshot listpending -C testchannel --peerAddress 127.0.0.1:22509 --tlsRootCertFile tls/cert.pem
```

You will see a response similar to:

```
Successfully got pending snapshot requests [970]

> ```

Reviewer: [1000]

Author: Sorry I meant for the submitrequest command above to be 970.

Reviewer: I believe that the 1000 would have covered more regular scenario. Because typically, if you specify a particular block number, that would be well ahead of the current height. Else, by the time the snapshot request is submitted the ledger height may change and the request will not be entertained. Also, in the `cancelrequest` command below also 1000 is being used. IMO, 1000 makes better sense in this example throughout.

> When a snapshot has been generated for a particular block height, the pending request for that block height will no longer appear in the list. You can also verify that a snapshot has been create successfully by looking at the peer logs. You will see a message similar to `Generated snapshot block-<block number>`.

Reviewer: I don't think that the actual log messages should go here as it's hard to maintain synergy between code and docs at this level of granularity. Instead, we can advise to look for the existence of the snapshot folder for the new snapshot.

To delete a snapshot request, simply exchange `submitrequest` with `cancelrequest`. For example:

```
peer snapshot cancelrequest -C testchannel -b 1000 --peerAddress 127.0.0.1:22509 --tlsRootCertFile tls/cert.pem
```

If you submit the `listpending` command again, the snapshot should no longer appear.

### Contents of a snapshot

> Snapshots are stored on a peer in a snapshot folder that is further organized by channel name and the ledger height when the snapshot was taken: `{snapshotsFolder}/{channelName}/{blockHeight}`.

Reviewer: >the ledger height

block number

>{snapshotsFolder}/{channelName}/{blockHeight}

{snapshotsFolder}/completed/{channelName}/{lastBlockNumberInSnapshot}

Also, better to mention that once the peer generates a snapshot in the above mentioned folder, the peer does not use that folder in any manner and it is safe to operate on these folders for the purpose of transferring the snapshots or compressing them using external tools

As mentioned above, the snapshot contains four folders:

* **Public state**

>   * This includes the latest state of all of the assets on the channel (for example, the current owner of a car).

Reviewer: >latest state of all of the assets on

public state part of the world state ("assets" is too narrow a term at this level)

Author: Ok I've tried to give a more precise example here: 

This includes the latest value of all of the keys on the channel. For example, the public state would show a car (a key) and its current owner (the value), but not any of its historical owners.

* **Private data hashes**

>   * Recall from our documentation on private data that while the actual transaction data of a private data transaction is not stored on the public ledger of the channel, hashes of the data are stored. This allows the private data to be verified against the hash, for example by an organization that is added to a private data collection.

Reviewer: Better to add one line in the beginning - "This includes the private data hashes part of the world state of the channel."

Author: Added a note about this in the limitations section.

Reviewer: The change in the limitations section LGTM. However, what I was not clear was whether you intend to add one liner here that explains what is included in the data under the heading (`Private data hashes`) or wanted to leave it with the heading being self explanatory. I am fine with either. But you want the latter, then the one liner can be removed from the previous heading  `Public state` as well for the consistency.

* **Transactions IDs**

>   * Starting with Fabric v2.0, entries in the block store index take the form of `TxID_BlkNum_TxNum`. This index can be filtered for the the transactions IDs that are added after the snapshot height by evaluating block number (`BlkNum`) portion of the index entry.

Reviewer: This consists of transaction ID that have been used till the last block in the snapshot

Reviewer: I guess that you missed tackling this in your latest commit. The current description under `Transactions IDs` is the implementation details that does not make sense here. So, like previous comment, if you intend to just keep the heading, this whole text can be removed and if you want to have one liner, then it can be replaced by "This consists of transaction ID that have been used till the last block in the snapshot"

* **Collection config history**

>   * This contains the private data collections the peer is a member of.

Reviewer: >private data collections the peer is a member of

history of the collection configurations for all the chaincodes. Recall from our documentation on private data that the collection configuration derive the endorsement and dissemination policies for private data collections

> 

Reviewer: Better to add the following sentence here...

Note that some of the above mentioned data may not be present in the snapshot. For instance, if none of the chaincodes in the channel use the private data, then the files for collection config history will not be generated.

Author: I'm not following you here. How can a channel contain private data unrelated to chaincodes being "used" (instantiated?).

Reviewer: I was suggesting that after describing the list of file types (corresponding to different data items) here, we can add one line saying that the file types explained here is a superset. So, if admins find some of these file types missing in their snapshots, they should not think that their snapshots are incomplete - as admins typically are not expected to know the application level details, they may simply look at the listing of file types here and try to match what they get on their snapshots.

> In addition, the snapshot contains several metadata files that help the data in the snapshot to be verified:

Reviewer: >several metadata files

two snapshot metadata files

* Channel name
* Height of the ledger at the time the snapshot was taken
* BlockHash for the last block in the snapshot
* Hash of the public state
* Hash of the private data hashes
* Hash of transaction IDs
* Hash of collection config history
* A representative hash of the snapshot, computed by combining all of the above
* `BlockCommitHash` for the last block in the snapshot

> * Compression algorithm used for the data files

Reviewer: * **_snapshot_signable_metadata.json**
<br>This snapshot metadata file is expected to be same across snapshots of a channel for a particular height. This file contains a JSON record with the following fields
  * channel_name
  * last_block_number
  * last_block_hash
  * previous_block_hash
  * state_db_type (the value of this field will be either of `CouchDB` and `SimpleKeyValueDB`. The latter will be the value if leveldb is used for statedb)
  * snapshot_files_raw_hashes (the value of this will be another JSON record that caontains the hashes of the files present for the above mentioned data items)

* **_snapshot_additional_metadata.json** 
<br>This another metadata file contains a JSON record with the following two fields
  * snapshot_hash (this is the hash of the file `_snapshot_signable_metadata.json` and can be treated as a hash of the snapshot)
  * last_block_commit_hash (if the snapshot generating peer is equipped to compute the block commit hashes)

Author: This is great, thank you!

## Joining a channel using a snapshot

> When joining a channel without using a snapshot, a command similar to `peer channel join --blockpath mychannel.block` is issued. When joining the peer to the channel using a snapshot, issue a command similar to:

Reviewer: >without using a snapshot
 
 with a genesis block

```
peer channel joinbysnapshot --snapshotPath <path to snapshot>
```

To verify that the peer has joined the channel successfully, issue a command similar to:

```
peer channel getinfo -c <name of channel joined by snapshot>
```

> Additionally, install at least one chaincode being used on the channel and issue a query. A successful return of data indicates that the peer has successfully joined using the snapshot. You can then install all of the chaincodes being used on the channel.

Reviewer: >install at least one chaincode being used on the channel,
 
if not already installed on the peer.

Also, add one more sentence here - If the snapshot is used by a new org, and the channel contains the definitions for the chaincodes that are defined using new lifecycle, the new org will need to approve the definition of such chaincodes before, the peer can be used for invoking these chaincode.

Author: Is it common for a chaincode to be installed on a peer before it joins a channel?

Reviewer: It may happen if the chaincode is already being used in one or more existing channels on the peer.

```

When a snapshot has been generated for a particular block height, the pending request for that block height will no longer appear in the list. You can also verify that a snapshot has been create successfully by looking at the peer logs or by checking the expected location of the snapshot, `{snapshotsFolder}/completed/{channelName}/{lastBlockNumberInSnapshot}`, for the snapshot.

To delete a snapshot request, simply exchange `submitrequest` with `cancelrequest`. For example:

```
peer snapshot cancelrequest -C testchannel -b 1000 --peerAddress 127.0.0.1:22509 --tlsRootCertFile tls/cert.pem
```

If you submit the `listpending` command again, the snapshot should no longer appear.

### Contents of a snapshot

Snapshots are stored on a peer in a snapshot folder that is further organized by channel name and the block number when the snapshot was taken: `{snapshotsFolder}/completed/{channelName}/{lastBlockNumberInSnapshot}`. Once the peer generates a snapshot in this folder, the peer does not use that folder for any purpose and it is safe to transfer or compress the snapshot using external tools.

As mentioned above, the snapshot contains four folders:

> 

Reviewer: Snapshot is a single folder that contains files for different data items listed below.

* **Collection config history**
  * This contains the private data collections the peer is a member of.

In addition, the snapshot contains two metadata files that help the data in the snapshot to be verified. This snapshot metadata file is expected to be same across snapshots of a channel for a particular height.

One of these files contains a JSON record called `_snapshot_signable_metadata.json` with the following fields:

* `channel_name`: the name of the channel.
* `last_block_number`: the block height when the snapshot was taken.
* `last_block_hash`: a hash of the block at the height the snapshot was taken.
* `previous_block_hash`: a hash of the block prior to the `last_block`.
* `state_db_type` (the value of this field will be either CouchDB or SimpleKeyValueDB (also known as LevelDB).
* `snapshot_files_raw_hashes`, is a JSON record that contains the hashes of the files present in `_snapshot_additional_metadata.json`.

> 

Reviewer: >hashes of the files present in `_snapshot_additional_metadata.json`.

hashes of the files present for the above mentioned data items - i.e., `Public state`, `Private data hashes`, `Transactions IDs`, and `Collection config history`

Reviewer: I think that I did not make myself clear enough in this. "hashes of the files present in _snapshot_additional_metadata.json" is a wrong/meaningless statement. So, I suggested a replacement.

Author: Ok I think I understand now. Changing the relevant bullet to: 

`snapshot_files_raw_hashes`, is a JSON record that contains the hashes of the files above

# Taking ledger snapshots and using them to join channels

> For a peer to process transactions on a channel, it must endorse and validate transactions consistent with the other peers on a channel. This includes the "world state", maintained in the state database, which represents the current value of all of the keys on the ledger (who owns a particular asset, for example) as of the most recently committed block. There are two ways for a peer to get a copy of the necessary ledger data.

Reviewer 3: The transition from the first to the second sentence got screwed up in the last changes. Probably my fault in the suggestions. @joealewine can you improve upon the wording?

Author: Yep I see it now.

# Taking ledger snapshots and using them to join channels

> For a peer to contribute to a channel, it must contain the minimum ledger data necessary to produce the same read/write sets as other peers when asked to endorse a transaction. This includes the "world state", maintained in the state database, which represents the current value of all of the keys on the ledger (who owns a particular car, for example) as of the most recently committed block. There are two ways for a peer to get a copy of the necessary ledger data.

Reviewer 3: It's important for both endorsement and validation.
And let's use the example of a generic 'asset' throughout instead of 'car', since we've shifted the tutorials to work against 'assets' instead of 'cars'.

```suggestion
For a peer to process transactions on a channel, it must contain the minimum ledger data necessary to endorse and validate transactions consistent with other peers on a channel. This includes the "world state", maintained in the state database, which represents the current value of all of the keys on the ledger (who owns a particular asset, for example) as of the most recently committed block. There are two ways for a peer to get a copy of the necessary ledger data.
```

2. Join the channel from a "snapshot", which contains the minimum ledger data as of a particular block number, without needing to pull and process the individual blocks.

While the first method represents a more comprehensive way of joining a channel, because of the size of established channels (which can reach many thousands of blocks), it can take a long time for peers to finish processing all the blocks already committed to the channel. Peers that join a channel this way must also store every block since the creation of the channel, increasing storage costs for an organization.

In this topic, we'll describe the process for joining a peer to a channel using a snapshot.

## Limitations

While joining a peer to a channel from a snapshot (also known as a "checkpoint") will take less time and saves on storage costs compared to processing and storing every block on the ledger, there are a few limitations to consider:

> * It is not possible for a peer that joins from a snapshot to query blocks that were committed before the ledger height of the snapshot. If the snapshot the peer uses to join a channel from was taken at block 1000, for example, none of the blocks between 1-999 can be queried. Applications attempting to query for the data in these blocks will have to target a peer that contains the relevant block. Because of this, it is likely that all organizations will want to keep at least one peer that has all of the historical data.

Reviewer 3: ```suggestion
* It is not possible for a peer that joins from a snapshot to query blocks (or transactions within blocks) that were committed before the ledger height of the snapshot. If the snapshot the peer uses to join a channel from was taken at block 1000, for example, none of the blocks between 0-999 can be queried. Applications attempting to query for the data in these blocks will have to target a peer that contains the relevant block. Because of this, it is likely that all organizations will want to keep at least one peer that has all of the historical data.
```

> * While endorsements will continue and queries can be submitted, all peers taking a snapshot at a particular height will not commit blocks on the channel while the snapshot is being generated. Because taking a snapshot is a resource-intensive operation, the peer might also be slow to endorse transactions or commit blocks on other channels. For these reasons, it is anticipated that snapshots will only be taken when necessary (for example, when a new peer is joining a channel or when organizations want to verify that no ledger forks have occurred). As long as every organization taking a snapshot has at least one peer that is not taking a snapshot, the operation of your organization and your channels should not be affected.

Reviewer 3: The heading paragraph mentions these are limitations of "joining a peer to a channel from a snapshot".
This limitation is not about joining, but about generation. Therefore, either make the heading paragraph more generic, or split the limitations into two lists, one for joining and one for generating snapshots.

> * While endorsements will continue and queries can be submitted, all peers taking a snapshot at a particular height will not commit blocks on the channel while the snapshot is being generated. Because taking a snapshot is a resource-intensive operation, the peer might also be slow to endorse transactions or commit blocks on other channels. For these reasons, it is anticipated that snapshots will only be taken when necessary (for example, when a new peer is joining a channel or when organizations want to verify that no ledger forks have occurred). As long as every organization taking a snapshot has at least one peer that is not taking a snapshot, the operation of your organization and your channels should not be affected.

Reviewer 3: "As long as every organization taking a snapshot has at least one peer that is not taking a snapshot, the operation of your organization and your channels should not be affected."

That is a stated a bit too strong. It depends how the application is written. If client applications are smart enough to target peers of higher block heights this is true. Endorsements and queries should target peers of higher height, so that the peers that are taking a snapshot and therefore slightly behind in block height are avoided.

* Because the private data between organizations in a channel is likely to be at least somewhat different, private data is not included in the snapshot (hashes of the private data are included, but not the data itself). Peers that join a channel using a snapshot will discover the collections it is a member of and pull the relevant private data from peers that are members of those collections directly.

> * The snapshot process is not meant to be a method to archive and prune the ledgers of peers that are already joined to the channel. Similarly, it is not meant as a method to take a full backup a peer, as peer configuration information, such as MSPs, are not included in a snapshot).

Reviewer 3: ```suggestion
* The snapshot process does not archive and prune the ledgers of peers that are already joined to the channel. Similarly, it is not meant as a method to take a full backup of a peer, as private data, and peer configuration information, such as MSPs, are not included in a snapshot).
```

> 

Reviewer 3: Another limitation:
It is not possible to use the reset, rollback, or rebuild-dbs commands on peers that have bootstrapped any channel from a snapshot since the peer wouldn't have all the block files required for the operations. Instead of these administrative commands, it is expected that peers that have been bootstrapped from snapshots can be entirely rebuilt from the same or newer snapshots.

Because of these limitations, the decision of whether a peer will join a channel from a snapshot depends on a few factors.

* **The current ledger height of the channel**. The smaller the ledger, the less useful it is to join a channel using a snapshot. Where the size of the ledger crosses a point where it makes sense to use a snapshot depends on your use case.
* **The importance of querying old data**. Some channels and use cases might call for querying old blocks regularly, while in other channels it might only be necessary to know the current state. This must be balanced by the server costs experienced by your organization.

## Overview

Snapshots can be used by organizations that already have peers on a channel or by organizations new to a channel. Whatever the use case, the process is largely the same.

> If the organization is new to the channel, the organizations currently joined to the channel must:

Reviewer 3: I wouldn't position these bullets as being for "new organizations" in this sentence, or within the bullets. The process is largely the same (as you say in the prior sentence).
I would document a common set of actions that are valid irrespective of the use case, and then use the following paragraph to mention that the process can be tailored for your use case.

> 1. **Schedule a snapshot**. These snapshots must be taken at **exactly the same ledger height** on each peer. This will allow the new peer organization to evaluate the snapshots to make sure they contain the same data. This ledger height must be equal or higher than the current ledger height. They cannot be taken from a lower ledger height. If you attempt to schedule a snapshot at a height lower than the current height you will get an error. Note that a peer that already used a snapshot to join a channel can also be used to take a snapshot.

Reviewer 3: Somehow work the word "future" in here, when talking about higher height.  Upon initial read people may be confused about how you can take a snapshot at a higher height than the peer is currently at (although it gets clarified in the next bullet).

Author: I couldn't get the word "future" in here, but I did attempt to clarify.

> 1. **Schedule a snapshot**. These snapshots must be taken at **exactly the same ledger height** on each peer. This will allow the new peer organization to evaluate the snapshots to make sure they contain the same data. This ledger height must be equal or higher than the current ledger height. They cannot be taken from a lower ledger height. If you attempt to schedule a snapshot at a height lower than the current height you will get an error. Note that a peer that already used a snapshot to join a channel can also be used to take a snapshot.

Reviewer 3: The text here implies that snapshots are taken as needed. But some networks may simply have a policy to snapshot peers at a regular cadence, for example every 10,000 blocks. That way consistent and recent snapshots are always available. It is worth mentioning as a consideration, either here or elsewhere. 

Note - it is not possible to schedule recurring snapshots (that may be a future enhancement). Each snapshot has to be scheduled independently. Although you can schedule as many future snapshots as you like.

2. **When the ledger height is reached, the snapshot is taken by the peer**. The snapshot is comprised of a folder that includes files that contain the public state, hashes of private state, transaction IDs, and the collection config history. A file containing metadata relating to these files is also included. For more information, check out [contents of a snapshot](#contents-of-a-snapshot).
2. **The snapshot is sent to the new organization**. This must be completed out of band. Because snapshot files are not compressed, it is likely that peer administrators will want to compress these files before sending them. In a typical scenario, the administrator will receive the snapshot from one of the existing organizations but will want to receive the snapshot metadata from more than one organizations in order to verify the snapshot received.

> The new organization will then:

Reviewer 3: Again, these steps are not specific to "new organizations".

> 1. **Evaluate the snapshots from the other organizations to ensure they contain the same data**. An administrator of the peer organization attempting to use the snapshot to join the peer to the channel should independently compute the hashes of the snapshot files and match these with the hashes present in the metadata file. In addition, the administrator may want to match the metadata files from more than one organization, depending on the trust model established by the network. In some scenarios, the administrator may want the administrators of other organizations to sign the metadata file for its records.

Reviewer 3: Again, even if you are not a new organization, you likely want to compare two snapshots from within your org before bootstrapping another peer from the snapshot.  So I'd put all the commentary about trust models and multiple organizations in the follow-on paragraph.

Author: I'm not following you. Shouldn't a discussion about the trust model of the network be contained in a paragraph about evaluating the snapshots?

> 2. **Join the peer to the channel using the snapshot**. When the peer has finished joining the channel using the snapshot, it will begin pulling private data according to the collections it is a member of.

Reviewer 3: ```suggestion
2. **Join the peer to the channel using the snapshot**. When the peer has finished joining the channel using the snapshot, it will begin pulling private data according to the collections it is a member of. It will also start committing blocks as normal, starting with any blocks greater than the snapshot height that are available from the ordering service.
```

3. **Verify the peer has joined the channel successfully**. For more information, check out [Joining a channel using a snapshot](#joining-a-channel-using-a-snapshot).

If an organization that is already joined to the channel wants to join a new peer using a snapshot, it might decide to skip the process of having other organizations take snapshots and evaluate them, though it is a best practice for an organization to periodically take snapshots of all its peers and compare them to ensure the no ledger forks have occurred. In that case, the organization can take a snapshot immediately and then use the snapshot to join the new peer to the channel.

> **Note: snapshots can also be used to verify that the public state between peers is identical (in other words, that no ledger fork has occurred), even if no new peer will use the snapshot to join the channel. This can be done by ensuring that the `snapshot_hash` in the file `_snapshot_additional_metadata.json` in the snapshots generated across peers should is the same.**

Reviewer 3: ```suggestion
**Note: snapshots can also be used to verify that the public state between peers is identical (in other words, that no ledger fork has occurred), even if no new peer will use the snapshot to join the channel. This can be done by ensuring that the `snapshot_hash` in the file `_snapshot_additional_metadata.json` in the snapshots generated across peers is the same.**
```

> 

Reviewer 3: Somewhere around here we should also mention a tip - when joining a peer from a snapshot it is a good practice to use a snapshot more recent than the latest channel config block height. This ensures that the peer will have the most recent channel configuration including the latest ordering service endpoints and CA certificates.

Author: Added this to the scheduling snapshots bullet.

```
peer snapshot submitrequest -C <name of channel> -b <ledger height where snapshot will be taken> --peerAddress <address of peer> --tlsRootCertFile <path to root certificate of the TLS CA>
```

For example:

```
peer snapshot submitrequest -C testchannel -b 1000 --peerAddress 127.0.0.1:22509 --tlsRootCertFile tls/cert.pem
```

**If you give a ledger height of `0`, the snapshot is taken immediately. This is useful for cases when an organization is interested in generating a snapshot that will be used by one of its own peers and does not intend to share the data with another organization. Do not take these "immediate" snapshots in cases when snapshots will be evaluated from multiple peers, as it increases the likelihood that the snapshots will be taken at different ledger heights.**

If the request is successful, you will see a `Snapshot request submitted successfully` message.

You can list the pending snapshots by issuing a command similar to:

```
peer snapshot listpending -C testchannel --peerAddress 127.0.0.1:22509 --tlsRootCertFile tls/cert.pem
```

You will see a response similar to:

```
Successfully got pending snapshot requests [1000]
```

> When a snapshot has been generated for a particular block height, the pending request for that block height will no longer appear in the list. You can also verify that a snapshot has been create successfully by looking at the peer logs or by checking the expected location of the snapshot, `{snapshotsFolder}/completed/{channelName}/{lastBlockNumberInSnapshot}`, for the snapshot.

Reviewer 3: We'll need to document where the snapshotFolder is. By default it is at peer.fileSystemPath. I believe it is configurable, although I'm not seeing the config option right now... @manish-sethi ?

```suggestion
When a snapshot has been generated for a particular block height, the pending request for that block height will no longer appear in the list. You can also verify that a snapshot has been created successfully by looking at the peer logs or by checking the expected location of the snapshot, `{snapshotsFolder}/completed/{channelName}/{lastBlockNumberInSnapshot}`, for the snapshot.
```

To delete a snapshot request, simply exchange `submitrequest` with `cancelrequest`. For example:

```
peer snapshot cancelrequest -C testchannel -b 1000 --peerAddress 127.0.0.1:22509 --tlsRootCertFile tls/cert.pem
```

If you submit the `listpending` command again, the snapshot should no longer appear.

### Contents of a snapshot

> Snapshots are stored on a peer in a snapshot folder that is further organized by channel name and the block number when the snapshot was taken: `{snapshotsFolder}/completed/{channelName}/{lastBlockNumberInSnapshot}`. Once the peer generates a snapshot in this folder, the peer does not use that folder for any purpose and it is safe to transfer or compress the snapshot using external tools.

Reviewer 3: ```suggestion
Snapshots are stored on a peer in a snapshot folder that is further organized by channel name and the block number when the snapshot was taken: `{snapshotsFolder}/completed/{channelName}/{lastBlockNumberInSnapshot}`. Once the peer generates a snapshot in this folder, the peer does not use that folder for any purpose and it is safe to compress and transfer the snapshot using external tools, and to delete it when no longer needed.
```

As mentioned above, the snapshot contains a folder with files for the different data items listed below:

* **Public state**

>   * This includes the latest value of all of the keys on the channel. For example, the public state would show a car (a key) and its current owner (the value), but not any of its historical owners.

Reviewer 3: ```suggestion
  * This includes the latest value of all of the keys on the channel. For example, the public state would show an asset (a key) and its current owner (the value), but not any of its historical owners.
```

* **Private data hashes**

>   * Recall from our documentation on private data that while the actual transaction data of a private data transaction is not stored on the public ledger of the channel, hashes of the data are stored. This allows the private data to be verified against the hash, for example by an organization that is added to a private data collection. These hashes are included in the snapshot so that new organizations can verify them against the private data they receive.

Reviewer: >These hashes are included in the snapshot so that new organizations can verify them against the private data they receive

I noticed this addition in the last commit. I would like to refrain from this as there is more to the hashes than this. In fact, hashes are present with every peer whether the peer is eligible for the private data or not.

Author: Well, I think we do need to say something here about what these hashes are for, even if it's not the whole truth. Maybe you can suggest an alternative here?

Reviewer: I had suggested a sentence in a [previous comment](https://github.com/hyperledger/fabric/pull/2043#discussion_r512322065). We have already mentioned before that the world state is one of the data items in the snapshot.

* **Transactions IDs**

>   * This consists of the transaction IDs that have been used in the channel until the last block in the snapshot.

Reviewer 3: ```suggestion
  * This consists of the transaction IDs that have been used in the channel until the last block in the snapshot. The transaction IDs are included so that peers can verify that a transaction ID is not later re-used for another transaction.
```

* **Collection config history**
  * This contains the history of the collection configurations for all chaincodes. Recall from our documentation on private data that the collection configuration derives the endorsement and dissemination policies for private data collections.

In addition, the snapshot contains two metadata files that help the data in the snapshot to be verified. This snapshot metadata file is expected to be same across snapshots of a channel for a particular height.

One of these files contains a JSON record called `_snapshot_signable_metadata.json` with the following fields:

* `channel_name`: the name of the channel.
* `last_block_number`: the block height when the snapshot was taken.
* `last_block_hash`: a hash of the block at the height the snapshot was taken.
* `previous_block_hash`: a hash of the block prior to the `last_block`.
* `state_db_type` (the value of this field will be either CouchDB or SimpleKeyValueDB (also known as LevelDB).
* `snapshot_files_raw_hashes`, is a JSON record that contains the hashes of the files above.

> This another metadata file is a JSON record with the following two fields:

Reviewer 3: Typo - "This another"

* `snapshot_hash`, the hash of the file `_snapshot_signable_metadata.json` and can be treated as a hash of the snapshot.
* `last_block_commit_hash`, which is included if the snapshot generating peer is equipped to compute the block commit hashes.

Note that the file types explained here is a superset of all of the files that might be included in a snapshot. If admins find some of these file types missing in their snapshots (for example, the collection config history) this is not mean the snapshot is incomplete. The channel might not have any collections.

## Joining a channel using a snapshot

When joining a channel using the genesis block, a command similar to `peer channel join --blockpath mychannel.block` is issued. When joining the peer to the channel using a snapshot, issue a command similar to:

```
peer channel joinbysnapshot --snapshotPath <path to snapshot>
```

To verify that the peer has joined the channel successfully, issue a command similar to:

```
peer channel getinfo -c <name of channel joined by snapshot>
```

> Additionally, if the peer has not already installed a chaincode being used on the channel, do so, and then issue a query. A successful return of data indicates that the peer has successfully joined using the snapshot. You can then install all of the chaincodes being used on the channel. If the snapshot is used by a new organization, and the channel contains the definitions for the chaincodes that are defined using the [new chaincode lifecycle](../chaincode_lifecycle.html), the new organization will need to approve the definition of these chaincodes can be invoked using the peer.

Reviewer 3: ```suggestion
Additionally, if the peer has not already installed a chaincode being used on the channel, do so, and then issue a query. A successful return of data indicates that the peer has successfully joined using the snapshot. You can then install all of the chaincodes being used on the channel. If the snapshot is used by a new organization, and the channel contains the definitions for the chaincodes that are defined using the [new chaincode lifecycle](../chaincode_lifecycle.html), the new organization will need to approve the definition of these chaincodes before they can be invoked on the peer.
```

Actually, this is not just for new organizations, you must approveformyorg for all chaincodes on each bootstrapped channel, as that is the step that associates a certain chaincode package with the channel definition on that peer.

Author: If the organization is already on the channel, it would have already gone through the "approveformyorg" step, right? This does seem like something only a "new" organization would have to do.

Reviewer 3: @joealewine You're right, it was late at night... only new orgs need to approveformyorg.

## Try it out

If you want to try out the ledger snapshotting process, you'll first need a network with a running channel. If you don't have a network, deploy the [test network](./test_network.html). This will create a network with two orgs, which both have a single peer, and an application channel.

Select the peer you want to use to take the snapshot and follow the instructions above to take the snapshot. Then locate the snapshot on the your peer and copy it somewhere else on your filesystem.

> After you have taken the snapshot and copied it, follow the [Adding an Org to a Channel](./channel_update_tutorial.html) to add a new org to your network and application channel. When you reach the section where you are asked to [Join Org3 to the Channel](./channel_update_tutorial.html#join-org3-to-the-channel), instead of the `peer channel join -b mychannel.block` command, substitute `peer channel joinbysnapshot --snapshotPath <path to snapshot>` using the path to the snapshot on your filesystem.

Reviewer: >After you have taken the snapshot and copied it, follow the [Adding an Org to a Channel]

The recommended way would be to take the snapshot after committing the new channel config that includes org3 (i.e., after completion of [Sign and Submit the Config Update] step) so, the snapshot includes the latest channel config. In other words, I would make the step for generating the snapshot as part of [Join Org3 to the Channel].
